In [4]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [5]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# YOUR CODE HERE!
df = pd.read_csv(
    Path("../data/lending_data.csv")
)
# Review the DataFrame
# YOUR CODE HERE!


df.head()

,loan_size,interest_rate,borrower_income,debt_to_income,num_of_accounts,derogatory_marks,total_debt,loan_status
0,10700.0,7.672,52800,0.431818,5,1,22800,0
1,8400.0,6.692,43600,0.311927,3,0,13600,0
2,9000.0,6.963,46100,0.349241,3,0,16100,0
3,10700.0,7.664,52700,0.430740,5,1,22700,0
4,10800.0,7.698,53000,0.433962,5,1,23000,0


### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [6]:
# Separate the data into labels and features
# Separate the y variable, the labels
# YOUR CODE HERE!]
y = df["loan_status"]

# Separate the X variable, the features
# YOUR CODE HERE!
X = df.drop(columns="loan_status")


In [7]:
# Review the y variable Series
# YOUR CODE HERE!

y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [8]:
# Review the X variable DataFrame
# YOUR CODE HERE!

X.head()

,loan_size,interest_rate,borrower_income,debt_to_income,num_of_accounts,derogatory_marks,total_debt
0,10700.0,7.672,52800,0.431818,5,1,22800
1,8400.0,6.692,43600,0.311927,3,0,13600
2,9000.0,6.963,46100,0.349241,3,0,16100
3,10700.0,7.664,52700,0.430740,5,1,22700
4,10800.0,7.698,53000,0.433962,5,1,23000


### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [9]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!

X_train, X_test, y_train, y_test = train_test_split(X, y, 
                                                    random_state=1,
                                                    stratify=y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [10]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
LogisticRegression_model = LogisticRegression(solver='lbfgs', random_state=1)
# Fit the model using training data
# YOUR CODE HERE!
Regression_model = LogisticRegression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [11]:
# Make a prediction using the testing data
# YOUR CODE HERE!
predictions = Regression_model.predict(X_test)

pd.DataFrame({"Prediction": predictions, "Actual": y_test}).reset_index(drop=True)

,Prediction,Actual
0,0,0
1,1,1
2,0,0
3,0,0
4,0,0
...,...,...
19379,0,0
19380,0,0
19381,0,0
19382,0,0


### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [12]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
confusion_matrix(y_test, predictions)

array([[18673,    86],
       [   32,   593]])

In [13]:
# Print the classification report for the model
# YOUR CODE HERE!
target_names = ["(0) Healthy Loan", "(1) High-Risk Loan"]
print(classification_report(y_test, predictions, target_names=target_names ))

                    precision    recall  f1-score   support

  (0) Healthy Loan       1.00      1.00      1.00     18759
(1) High-Risk Loan       0.87      0.95      0.91       625

          accuracy                           0.99     19384
         macro avg       0.94      0.97      0.95     19384
      weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** WRITE YOUR ANSWER HERE!
For the '0' (healthy loans) the precision is 100% , the recall is also a 100% where the actual healthy loans were correctly classified. 
for the '1' (high-risk loans) the precision is 87%, the recall is 95% which identified the actual high-risk loans correctly. 

The model has a high prediction of both (healthy loans) and (high_risk loans), it had a 100% accuracy for (healthy loans) and a slightly lower accuracy for the (high-risk loans) with a precision of 87% accuracy. 

---